### ISSUE TITLE
Building for Windows on macOS creates broken builds.

### ISSUE THREAD

**Issue by nmggithub** [electron-userland/electron-builder#8334]
* **Electron-Builder Version**: [25.0.0-alpha.13](https://www.npmjs.com/package/electron-builder/v/25.0.0-alpha.13)

* **Node Version**: v20.11.1

* **Electron Version**: 31.1.0
* **Electron Type (current, beta, nightly)**:

* **Electron Updater**: 6.3.0-alpha.6

* **Target**: Windows

I am having the same issue as the user in this StackOverflow thread:
https://stackoverflow.com/questions/74060787/electron-application-built-on-macos-for-windows-is-not-working-on-installation

It appears that builds for Windows on macOS do not actually work. The installer exists. But when I go and run the installer on Windows, it creates a nearly-empty folder in `%LocalAppData%\Programs`. The only file contained within is the uninstaller executable. Why does this happen?

**Comment by nmggithub** [comment]
I have just confirmed that Linux builds of the installers still work fine. macOS seems to be the common factor here. Please note that, because the macOS-built installer is broken, this causes auto-update to basically kill the application and delete all its files. This is not good user experience.

**Comment by mmaietta** [comment]
Is this a new issue in the alpha releases? Or also existing in [24.13.3](https://www.npmjs.com/package/electron-builder/v/24.13.3)?

**Comment by nmggithub** [comment]
It exists on the latest stable building, 24.13.13, yes. And, if that StackOverflow post is to be believed, it's been a thing since at least 23.6.0, but not 23.3.3.

**Comment by mmaietta** [comment]
Can you please create a minimum reproducible repo for this? I'm doubting this is a widespread issue as this is the first I've heard of it and I never have run into it with my local projects either.

**Comment by nmggithub** [comment]
I'll see what I can do. The app I am currently working with with it a proprietary app, so I don't feel comfortable sharing the code. I'm a bit busy with work, but when I find the time, I'll try and put together a minimal example with a new app.

**Comment by beyondkmp** [comment]
```
{
  "name": "ceaseless-volleyball-agree-oy9zj",
  "productName": "ceaseless-volleyball-agree-oy9zj",
  "description": "My Electron application description",
  "keywords": [],
  "main": "./main.js",
  "version": "1.0.0",
  "author": "beyondkmp",
  "scripts": {
    "start": "electron .",
    "dist": "electron-builder --win"
  },
  "build": {
    "appId": "electron-blog-example",
    "win": {
      "target": "NSIS"
    }
  },
  "devDependencies": {
    "electron-builder": "24.13.3",
    "electron": "27.2.0"
  }
}
```
I cannot reproduce this bug in macOS intel.

**Comment by beyondkmp** [comment]
@nmggithub Can you only package the Windows ARM version on a Mac ARM? I get an error when trying to package the Windows x64 version on macOS ARM.

**Comment by nmggithub** [comment]
> @nmggithub Can you only package the Windows ARM version on a Mac ARM? I get an error when trying to package the Windows x64 version on macOS ARM.

I don't have a config key set for x64 specifically. Just NSIS. It builds a Setup.exe just fine, but when I actually run it on a Windows machine, the result is an empty program folder.

**Comment by praisethemoon** [comment]
> I'm running it on macOS ARM. That could be part of it.

Happened to me too, I think targeting Win from Mac M series generates ARM windows builds.

**Comment by nmggithub** [comment]
> > I'm running it on macOS ARM. That could be part of it.
> 
> Happened to me too, I think targeting Win from Mac M series generates ARM windows builds.

If they are ARM Windows builds wouldn't the installer just not run at all? Why does it run and then just delete everything? Unless the installer is meant to be cross-architecture and then the code of our applications target ARM.

**Comment by nmggithub** [comment]
This is still happening. I have to build on an entirely different machine than my dev machine every time I want to deploy. It's not ideal.

**Comment by mmaietta** [comment]
@nmggithub can you provide a minimum reproducible repo for me to test locally with? I have a Mac arm64

**Comment by nmggithub** [comment]
> @nmggithub can you provide a minimum reproducible repo for me to test locally with? I have a Mac arm64

Just to confirm, are you unable to replicate it yourself with just a new empty project? As mentioned, the project I am working on is proprietary code, so I would have to first untangle that mess. Seeing as there are others who are facing this issue, this seems like a general thing that would happen regardless of any particulars about the project (other than building on a Mac). I could be wrong though.

**Comment by mmaietta** [comment]
Correct. On my test project, I built for windows x64 on my arm64 mac and installed it on my arm64 windows VM and my Windows PC. App runs correctly on both. Built using electron-builder 25.1.3 using my local test project based off electron-quick-start 
<img width="1118" alt="Screenshot 2024-10-01 at 8 44 42 AM" src="https://github.com/user-attachments/assets/74abb4ab-b97e-47be-9725-723cf416eeb6">

**Comment by nmggithub** [comment]
> Correct. On my test project, I built for windows x64 on my arm64 mac and installed it on my arm64 windows VM and my x64 Windows PC. App runs correctly on both. Built using electron-builder 25.1.3 using my local test project based off electron-quick-start 
> <img width="1118" alt="Screenshot 2024-10-01 at 8 44 42 AM" src="https://github.com/user-attachments/assets/74abb4ab-b97e-47be-9725-723cf416eeb6">
> ![image](https://github.com/user-attachments/assets/7bbc7bd4-ebf8-4cc8-9beb-b9e951279d38)
> 

Interesting. I'll have to see what is different on my end. I have a lot going on, but I'll try to be back at this thread soon with a minimal example.

**Comment by ctapu4ok** [comment]
I have the same problem, but I'm running on windows 64 for windows 64. 
Before that, the application (Setup) was assembled normally, and suddenly it suddenly stopped assembling at

When unpacked, the application works fine, but if you create an installer, then after installation it creates only uninstall.exe
So far, all we've been able to figure out is that it creates a broken archive in Setup.exe

![Image](https://github.com/user-attachments/assets/645717fe-3aab-4a47-9bfc-8460849e7649)

![Image](https://github.com/user-attachments/assets/62ebf2dd-e155-4fe3-bb7d-d21d1ccfd1fc)

![Image](https://github.com/user-attachments/assets/42b87a97-b409-4452-8a4d-4f3480f0acfc)

**Comment by nmggithub** [comment]
> To anyone running into this, the problem is that electron-builder builds for the architecture of the local machine, by default. In my case I was building on an apple M1 chip which would create `arm64` executables both for mac AND win, which caused the windows executable to not run at all. This is likely to be the issue with the OP as well since their build command does not include the `--x86` flag but [@mmaietta](https://github.com/mmaietta) 's does
> 
> Simply add `--x86` to the build command to build a x86 executable to fix this.

I'll take a look at this on my end and see if it resolves my issue. If it does, that would be fantastic!

**Comment by nmggithub** [comment]
> To anyone running into this, the problem is that electron-builder builds for the architecture of the local machine, by default. In my case I was building on an apple M1 chip which would create `arm64` executables both for mac AND win, which caused the windows executable to not run at all. This is likely to be the issue with the OP as well since their build command does not include the `--x64` flag but [@mmaietta](https://github.com/mmaietta) 's does
> 
> Simply add `--x64` to the build command to build a x64 executable to fix this.

This appears to have been the answer!

@mmaietta it looks like I missed the flag in your screenshot. Thank you @pedro-pedrosa for showing me the way!

I will close this issue, as I believe it is solved now.
